Handle soft proxies for custom assert classes in OSGi bundles (#1979)

bjhargrave opened the pull request:
When using a composite class loader for defining soft proxies when the
assert class is from a different class loader than the AssertJ classes,
we now use the composite class loader as the ClassLoadingStrategy for
ByteBuddy. This is necessary for Java 9+ to avoid the
ClassLoadingStrategy.UsingLookup default strategy used by AssertJ for
Java 9+. That strategy always defines the proxy classes in the class
loader defining the assert class ignoring the specified class loader.
This change is also beneficial for Java 8 as it avoids the need to
reflectively call the composite class loader to define the proxy
classes.

We refactor the CompositeClassLoader into the
ClassLoadingStrategyFactory class so that we can centralize the logic
for class loader and class loading strategy determination. Assumptions
is updated to use these changes which prepares it for a future where it
could support defining custom assumptions.

Finally, we update the class loader used for the ByteBuddy class caches
to use the class loader of the assert class rather than AssertJ's class
loader as the key. This is important in the OSGi case where multiple
bundles could define custom assertions with the same class name and we
must allow for unique proxy class generation for each bundle.

Fixes https://github.com/joel-costigliola/assertj-core/issues/1979
Fixes #1888

joel-costigliola commented on pom.xml:
```diff
@@ -96,6 +96,35 @@
+    false positive from the enforcer plugin. The Bnd 5.2.0
+    plugins don't have this problem, so these dependencies
+    can be removed when updating bnd.version to 5.2.0 and
+    setting them to use includeDependencyManagement=true.
```
does this comment apply for both `includeDependencyManagement` in bnd-testing-maven-plugin and bnd-resolver-maven-plugin?

bjhargrave replied:
Yes. I will improve the comment to be more specific.

joel-costigliola commented on src/test/java/org/assertj/core/osgi/soft/SimpleTest.java:
```diff
@@ -0,0 +1,33 @@
+
+  @Test
+  void simple_failure() {
+    assertThatCode(() -> assertThat("A String").isNull()).isNotNull();
```
what is being tested exactly here? that the `AssertionError` thrown by `assertThat("A String").isNull()` is properly captured?
would be clearer to replace `isNotNull` by `isInstanceOf(AssertionError.class)`.

bjhargrave replied:
I will make this improvement.

joel-costigliola commented on src/test/java/org/assertj/core/osgi/soft/CustomSoftAssertionTest.java:
```diff
@@ -0,0 +1,75 @@
+    map.put("key2", "value2");
+
+    TestSoftAssertions softly = new TestSoftAssertions();
+    softly.assertThat(map).containsKeys("key1", "key1").containsValues("value1", "value2");
```
typo: `"key1", "key2"`

bjhargrave replied:
Oops! I will fix.

kriegfrj replied:
Maybe this is a consideration for an enhancement- containsKeys() could throw an IllegalArgumentException if you pass in a duplicate? Would help catch accidental errors ingress this.

joel-costigliola reviewed (commented):
@bjhargrave looks good, just a few comments to address

joel-costigliola commented on src/main/java/org/assertj/core/api/Assumptions.java:
```diff
@@ -1277,18 +1278,19 @@ public static OptionalLongAssert assumeThat(OptionalLong actual) {
                      .intercept(ASSUMPTION)
                      .make()
-                     .load(Assumptions.class.getClassLoader(), classLoadingStrategy(assertionType))
+                     .load(strategy.getClassLoader(), strategy.getClassLoadingStrategy())
```
@bjhargrave can you add a test for these changes?

bjhargrave replied:
Done given the limitations that one cannot assume on a custom assertion class.

joel-costigliola reviewed (changes requested):
We will release 3.17.1 when this PR is integrated.
Only need a test to cover the change on the `Assumptions` class

bjhargrave commented:
> I'm moving this to 3.18.0 as there are build failures:
> 
> ```
> [ERROR] Resolution failed. Capabilities satisfying the following requirements could not be found:
>     [<<INITIAL>>]
>       ⇒ osgi.identity: (osgi.identity=assertj-core-tests)
>           ⇒ [assertj-core-tests version=3.17.1.202008300422]
>               ⇒ osgi.wiring.package: (osgi.wiring.package=org.assertj.core.test)
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  01:01 min
> [INFO] Finished at: 2020-08-30T16:23:19+12:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal biz.aQute.bnd:bnd-resolver-maven-plugin:5.1.2:resolve (osgi-integration-resolving) on project assertj-core: Unable to resolve <<INITIAL>>: missing requirement osgi.identity;filter:='(osgi.identity=assertj-core-tests)' [caused by: Unable to resolve assertj-core-tests version=3.17.1.202008300422: missing requirement osgi.wiring.package;filter:='(osgi.wiring.package=org.assertj.core.test)';osgi.wiring.package='org.assertj.core.test'] -> [Help 1]
> ```
>
> The issue comes from the snapshot version resolved to the minute, when the build started the version is `3.17.1.202008300422` but when the osgi tests are executed time was `202008300423`.

I don't know what step you took to get this issue. I can't figure out how to do it. The timestamp on the Bundle-Version is likely not the issue (I will fix that in the pom so it just says SNAPSHOT). The issue is that the assertj-core-tests bundle wants to import the org.assertj.core.test package which is exported by no one. In my build (I rebased on HEAD), I see:

```
bnd print --impexp /Users/hargrave/git/assertj-core/target/assertj-core-3.17.2-SNAPSHOT-tests.jar
[IMPEXP]

Import-Package
  org.assertj.core.api                   
  org.junit.jupiter.api                  {version=[5.6,6)}
```

Which is what is expected.

bjhargrave commented:
> Only need a test to cover the change on the `Assumptions` class

I added a simple Assumptions test case and updated the pom to use SNAPSHOT instead of timestamp in Bundle-Version for snapshot builds.

Since Assumptions does not have any support for assumptions on custom assertion classes, I could not test assumptions with a custom assertion class from an OSGi bundle.

joel-costigliola commented:
The build is still failing locally, unit tests pass but a similar error occurs:
```
[INFO] Bnd inferred -runee: JavaSE-1.8
[ERROR] Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=assertj-core-tests)
          ⇒ [assertj-core-tests version=3.17.2.SNAPSHOT]
              ⇒ osgi.wiring.package: (osgi.wiring.package=org.assertj.core.test)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  52.258 s
[INFO] Finished at: 2020-09-01T23:07:49+12:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal biz.aQute.bnd:bnd-resolver-maven-plugin:5.1.2:resolve (osgi-integration-resolving) on project assertj-core: Unable to resolve <<INITIAL>>: missing requirement osgi.identity;filter:='(osgi.identity=assertj-core-tests)' [caused by: Unable to resolve assertj-core-tests version=3.17.2.SNAPSHOT: missing requirement osgi.wiring.package;filter:='(osgi.wiring.package=org.assertj.core.test)';osgi.wiring.package='org.assertj.core.test'] -> [Help 1]
```
I'm a bit puzzled as the CI build is passing, I have tried with java 8 and java 11 for the same result.
My environment looks like:
```
23:07 $ ./mvnw -version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /home/joel/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 1.8.0_242, vendor: Oracle Corporation, runtime: /home/joel/.sdkman/candidates/java/8.0.242-open/jre
Default locale: en_NZ, platform encoding: UTF-8
OS name: "linux", version: "5.3.0-40-generic", arch: "amd64", family: "unix"
```
I'm on linux mint 19.3 which is based on ubuntu.

bjhargrave commented:
> The build is still failing locally

Perhaps your local .m2 cache is corrupt? Try `./mvnw -V --no-transfer-progress -e clean verify javadoc:javadoc -Dmaven.repo.local=./m2` to use the local `./m2` folder as a new, fresh m2 cache. This works fine for me on my mac.

```
./mvnw -V --no-transfer-progress -e clean verify javadoc:javadoc -Dmaven.repo.local=./m2 
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /Users/hargrave/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 11.0.8, vendor: Azul Systems, Inc., runtime: /Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.15.6", arch: "x86_64", family: "mac"
```

bjhargrave commented:
If a new m2 cache does not help, can you send me the built assertj-core-tests artifact so I can examine it to see why it has this additional import package?

joel-costigliola commented:
The new m2 cache did not fix the issue (same error).

I have renamed the jar extension to zip as github does not allow attaching jars: [assertj-core-3.17.2-SNAPSHOT-tests.zip](https://github.com/joel-costigliola/assertj-core/files/5161543/assertj-core-3.17.2-SNAPSHOT-tests.zip).
Here are the logs: [log.txt](https://github.com/joel-costigliola/assertj-core/files/5161549/log.txt)

bjhargrave commented:
You do not appear to be building the latest commits from this PR as your built test jar does not contain all the classes from this PR. I have made some updates and push updated commits since the original push. So please recollect the commits from this PR for your local build.

Here is the content of the test jar I build locally:

```
$ bnd print --full --list target/assertj-core-3.17.2-SNAPSHOT-tests.jar
[MANIFEST]

Build-Jdk-Spec                          11
Bundle-Description                      Rich and fluent assertions for testing for Java
Bundle-Developers                       joel-costigliola;email="joel.costigliola at gmail.com";name="Joel Costigliola";roles="Owner,Developer"
Bundle-DocURL                           http://assertj.org/assertj-core
Bundle-License                          "Apache License, Version 2.0";link="http://www.apache.org/licenses/LICENSE-2.0.txt"
Bundle-ManifestVersion                  2
Bundle-Name                             AssertJ fluent assertions
Bundle-SCM                              connection="scm:git:git@github.com:joel-costigliola/assertj-core.git"
                                        developer-connection="scm:git:git@github.com:joel-costigliola/assertj-core.git"
                                        tag=HEAD
                                        url="git@github.com:joel-costigliola/assertj-core"
Bundle-SymbolicName                     assertj-core-tests
Bundle-Vendor                           AssertJ
Bundle-Version                          3.17.2.SNAPSHOT
Created-By                              Maven Archiver 3.4.0
Import-Package                          org.assertj.core.api
                                        org.assertj.core.presentation
                                        org.junit.jupiter.api;version="[5.6,6)"
                                        org.opentest4j;version="[1.2,2)"
Manifest-Version                        1.0
Require-Capability                      osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Test-Cases                              org.assertj.core.osgi.AssumptionsTest,org.assertj.core.osgi.SimpleTest,org.assertj.core.osgi.soft.CustomSoftAssertionTest


[IMPEXP]

Import-Package
  org.assertj.core.api                   
  org.assertj.core.presentation          
  org.junit.jupiter.api                  {version=[5.6,6)}
  org.opentest4j                         {version=[1.2,2)}


[CAPABILITIES]

Require-Capability
  osgi.ee                                {filter:=(&(osgi.ee=JavaSE)(version=1.8))}

[COMPONENTS]



[METATYPE]



[API USES]


[USES]

org.assertj.core.osgi                   org.assertj.core.api
                                        org.assertj.core.presentation
                                        org.junit.jupiter.api
                                        org.opentest4j
org.assertj.core.osgi.soft              org.assertj.core.api
                                        org.junit.jupiter.api


[USEDBY]

org.assertj.core.api                    org.assertj.core.osgi
                                        org.assertj.core.osgi.soft
org.assertj.core.presentation           org.assertj.core.osgi
org.junit.jupiter.api                   org.assertj.core.osgi
                                        org.assertj.core.osgi.soft
org.opentest4j                          org.assertj.core.osgi


[LIST]

META-INF
  MANIFEST.MF
META-INF/maven
META-INF/maven <no contents>
META-INF/maven/org.assertj
META-INF/maven/org.assertj <no contents>
META-INF/maven/org.assertj/assertj-core
  pom.properties
  pom.xml
org
org <no contents>
org/assertj
org/assertj <no contents>
org/assertj/core
org/assertj/core <no contents>
org/assertj/core/osgi
  AssumptionsTest.class
  SimpleTest.class
org/assertj/core/osgi/soft
  CustomSoftAssertionTest$TestProxyableMapAssert.class
  CustomSoftAssertionTest$TestSoftAssertions.class
  CustomSoftAssertionTest.class
```

Here is the content of the jar you attached above:

```
$ bnd print --full --list assertj-core-3.17.2-SNAPSHOT-tests.zip
[MANIFEST]

Build-Jdk-Spec                          11
Bundle-Description                      Rich and fluent assertions for testing for Java
Bundle-Developers                       joel-costigliola;email="joel.costigliola at gmail.com";name="Joel Costigliola";roles="Owner,Developer"
Bundle-DocURL                           http://assertj.org/assertj-core
Bundle-License                          "Apache License, Version 2.0";link="http://www.apache.org/licenses/LICENSE-2.0.txt"
Bundle-ManifestVersion                  2
Bundle-Name                             AssertJ fluent assertions
Bundle-SCM                              connection="scm:git:git@github.com:joel-costigliola/assertj-core.git"
                                        developer-connection="scm:git:git@github.com:joel-costigliola/assertj-core.git"
                                        tag=HEAD
                                        url="git@github.com:joel-costigliola/assertj-core"
Bundle-SymbolicName                     assertj-core-tests
Bundle-Vendor                           AssertJ
Bundle-Version                          3.17.2.SNAPSHOT
Created-By                              Maven Archiver 3.4.0
Import-Package                          org.assertj.core.api
                                        org.assertj.core.data
                                        org.assertj.core.test
                                        org.assertj.core.util
                                        org.junit.jupiter.api;version="[5.6,6)"
Manifest-Version                        1.0
Require-Capability                      osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Test-Cases                              org.assertj.core.osgi.SimpleTest,org.assertj.core.osgi.soft.CustomSoftAssertionTest


[IMPEXP]

Import-Package
  org.assertj.core.api                   
  org.assertj.core.data                  
  org.assertj.core.test                  
  org.assertj.core.util                  
  org.junit.jupiter.api                  {version=[5.6,6)}


[CAPABILITIES]

Require-Capability
  osgi.ee                                {filter:=(&(osgi.ee=JavaSE)(version=1.8))}

[COMPONENTS]



[METATYPE]



[API USES]


[USES]

org.assertj.core.osgi                   org.assertj.core.api
                                        org.assertj.core.util
                                        org.junit.jupiter.api
org.assertj.core.osgi.soft              org.assertj.core.api
                                        org.assertj.core.data
                                        org.assertj.core.test
                                        org.junit.jupiter.api


[USEDBY]

org.assertj.core.api                    org.assertj.core.osgi
                                        org.assertj.core.osgi.soft
org.assertj.core.data                   org.assertj.core.osgi.soft
org.assertj.core.test                   org.assertj.core.osgi.soft
org.assertj.core.util                   org.assertj.core.osgi
org.junit.jupiter.api                   org.assertj.core.osgi
                                        org.assertj.core.osgi.soft


[LIST]

META-INF
  MANIFEST.MF
META-INF/maven
META-INF/maven <no contents>
META-INF/maven/org.assertj
META-INF/maven/org.assertj <no contents>
META-INF/maven/org.assertj/assertj-core
  pom.properties
  pom.xml
org
org <no contents>
org/assertj
org/assertj <no contents>
org/assertj/core
org/assertj/core <no contents>
org/assertj/core/osgi
  SimpleTest.class
org/assertj/core/osgi/soft
  CustomSoftAssertionTest$TestProxyableMapAssert.class
  CustomSoftAssertionTest$TestSoftAssertions.class
  CustomSoftAssertionTest.class
```

PS. The `bnd` command is a shell script which runs `java -jar biz.aQute.bnd-5.1.2.jar`. The `biz.aQute.bnd-5.1.2.jar` is available in maven central. I installed the `bnd` command from brew for my mac. But you use linux...

joel-costigliola commented:
@bjhargrave apologies :facepalm:   

When I integrate PRs, I usually rebase them on `main` locally (this is why the commit id differ), I refactored the tests a bit and used `org.assertj.core.test.Maps.mapOf` to initialize the maps like this:
```
    TestSoftAssertions softly = new TestSoftAssertions();
    Map<String, String> map = mapOf(entry("key1", "value1"), entry("key2", "value2"));
```  
This I guess introduced a dependency on `org.assertj.core.test` module and failed the build with the error I reported.

All good now.

joel-costigliola commented:
Integrated, thanks @bjhargrave!

bjhargrave commented:
>  I refactored the tests a bit and used `org.assertj.core.test.Maps.mapOf` to initialize the maps like this:

For the osgi integrations tests, they are packaged in the assertj-core-tests bundle and only have access to the public API from the assertj-core bundle. So they must only depend upon the public AssertJ API and not any helper classes from the tests src folder.
